mixnet/node/src/lib.rs

```diff
@@ -125,31 +138,43 @@ impl MixnetNode {
         tokio::time::sleep(delay.to_duration()).await;
 
         Self::forward(
+            cache,
             Body::new_sphinx(packet),
             NymNodeRoutingAddress::try_from(next_node_addr)?,
         )
         .await
     }
 
     async fn forward_payload_to_destination(
+        cache: ConnectionPool,
         payload: Payload,
         destination_addr: DestinationAddressBytes,
     ) -> Result<(), Box<dyn Error>> {
         tracing::debug!("Forwarding final payload to destination mixnode");
 
         Self::forward(
+            cache,
             Body::new_final_payload(payload),
             NymNodeRoutingAddress::try_from_bytes(&destination_addr.as_bytes())?,
         )
         .await
     }
 
-    async fn forward(body: Body, to: NymNodeRoutingAddress) -> Result<(), Box<dyn Error>> {
+    async fn forward(
+        cache: ConnectionPool,
+        body: Body,
+        to: NymNodeRoutingAddress,
+    ) -> Result<(), Box<dyn Error>> {
         let addr = SocketAddr::try_from(to)?;
+        if let Some(addr) = cache.get(&addr) {
+            let mut stream = addr.lock().await;
+            body.write(&mut *stream).await?;
+            return Ok(());
+        }
 
         let mut socket = TcpStream::connect(addr).await?;
         body.write(&mut socket).await?;
-
+        cache.insert(addr, Arc::new(tokio::sync::Mutex::new(socket)));
```

Comment:
we can probably use the `get_or_init` that you just added.

Reply:
The trade-off about using `get_or_init` here is we have to call `lock`.

```rust
let socket: Arc<Mutex<TcpStream>> = cache.get_or_init(addr).await;
let mut mu = socket.lock().await;
body.write(&mut socket).await?;
```

Reply:
Yes, we have tradeoff. If we don't want to allow the line 175 to be executed concurrently for the same host, we should use `get_or_init`. In the extreme case, this will cause the 'ran out of fd` error.

Reply:
As long as we decide to use the map with mutex, it means we're gonna take the overhead of lock contention whenever we handle connections. If so, it's better to have the consistent behaviour everywhere, imho.